python/ql/lib/semmle/python/frameworks/Asyncpg.qll

```diff
@@ -0,0 +1,192 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `asyncpg` PyPI package.
+ * See https://magicstack.github.io/asyncpg/.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+
+/** Provides models for the `asyncpg` PyPI package. */
+private module Asyncpg {
+  /** A `ConectionPool` is created when the result of `asyncpg.create_pool()` is awaited. */
+  API::Node connectionPool() {
+    result = API::moduleImport("asyncpg").getMember("create_pool").getReturn().getAwaited()
+  }
+
+  /**
+   * A `Connection` is created when
+   * - the result of `asyncpg.connect()` is awaited.
+   * - the result of calling `aquire` on a c=`ConnectionPool` is awaited.
+   */
+  API::Node connection() {
+    result = API::moduleImport("asyncpg").getMember("connect").getReturn().getAwaited()
+    or
+    result = connectionPool().getMember("acquire").getReturn().getAwaited()
+  }
+
+  /** Reverse lookup of the query argument name for a query method. */
+  private string queryMethodName(string queryArg) {
+    result in ["copy_from_query", "execute", "fetch", "fetchrow", "fetchval"] and
+    queryArg = "query"
+    or
+    result = "executemany" and
+    queryArg = "command"
+  }
```

Comment:
I'm not sure how I feel about this reverse-lookup predicate. I think I find it a bit "too clever", and impairing readability a bit. I think I would prefer a more verbose approach such as:


```codeql
  /** `Connection`s and `ConnectionPool`s provide some methods that execute SQL. */
  class SqlExecutionOnConnection extends SqlExecution::Range, DataFlow::MethodCallNode {
    string methodName;

    SqlExecutionOnConnection() {
      methodName in ["copy_from_query", "execute", "fetch", "fetchrow", "fetchval", "executemany"] and
      this.calls([connectionPool().getAUse(), connection().getAUse()], methodName)
    }

    override DataFlow::Node getSql() {
      methodName in ["copy_from_query", "execute", "fetch", "fetchrow", "fetchval"] and
      result in [this.getArg(0), "query"]
      or
      methodName = "executemany" and
      result in [this.getArg(0), "command"]
    }
  }
```

Would you be happy about such a rewrite?